From 00ddfa4919020dfe316e7d949fc9b2a03911886b Mon Sep 17 00:00:00 2001
From: Paul Naoumenko <paulx.naoumenko@intel.com>
Date: Fri, 25 Jan 2013 11:34:35 +0100
Subject: [PORT FROM MAIN] wifi: Cancel sched scan and scan before enabling pno, enable scan after disabling pno

BZ: 60560

Cancel any pending regular scan or supplicant sched scan (not driver's)
for power consumption optimization.
Reschedule a scan when pno is disabled.

Feature   : Wlan
Category  : Improvement
Purpose   : AOSP+
Origin    : Upstream
Upstream-candidate: yes

Change-Id: I3763bc3736e5abcb90d83d981d60fc05f9599c6a
Orig-Change-Id: Ieb3f7fb129674c12ae54b11362b086f4f341af3d
Signed-off-by: Paul Naoumenko <paulx.naoumenko@intel.com>
---
 wpa_supplicant/ctrl_iface.c |   10 ++++++++--
 1 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/wpa_supplicant/ctrl_iface.c b/wpa_supplicant/ctrl_iface.c
index ad0392f..2d75b83 100644
--- a/wpa_supplicant/ctrl_iface.c
+++ b/wpa_supplicant/ctrl_iface.c
@@ -370,10 +370,16 @@ static int wpa_supplicant_ctrl_iface_set(struct wpa_supplicant *wpa_s,
 		wpa_tdls_enable(wpa_s->wpa, !disabled);
 #endif /* CONFIG_TDLS */
 	} else if (os_strcasecmp(cmd, "pno") == 0) {
-		if (atoi(value))
+		if (atoi(value)) {
+			wpa_supplicant_cancel_sched_scan(wpa_s);
+			wpa_supplicant_cancel_scan(wpa_s);
 			ret = pno_start(wpa_s);
-		else
+		}
+		else {
 			ret = pno_stop(wpa_s);
+			if(wpa_supplicant_req_sched_scan(wpa_s))
+				wpa_supplicant_req_scan(wpa_s, 0, 0);
+		}
 	} else if (os_strcasecmp(cmd, "radio_disabled") == 0) {
 		int disabled = atoi(value);
 		if (wpa_drv_radio_disable(wpa_s, disabled) < 0)
-- 
1.7.4.1

